Windows: Support for Clingo and dependencies

johnwparent opened the pull request:
This PR ports Clingo, Bison, re2c, and other related components of Spack to function on Windows.

PR is currently draft pending discussion on Bison package implementation.

Clingo and re2c prove simple to port, however Bison has no official support for use of MSVC compilers, and so the port is more complex.

A project [winflexbison](https://github.com/lexxmark/winflexbison) vendors support for both Bison and Flex. This is currently introduced into Spack as a package named `winbison` and only support for the Bison components of the project is added.
Alternatively, this package could be introduced as a provider for Bison and Flex, packages which would then need to be made virtual, and exposed providers on a per platform basis. 

An additional alternative that has been proposed by @scheibelp is the creation of separate repositories for packages where support across Windows/Unix based platforms warrants the creation of a separate codebase.

johnwparent committed:
Clingo win support (#27300)

* First pass clingo windows support

* Add clingo-bootstrap

Re enable Clingo default on Windows

Refactor re2c to support windows

non virtual

proper py-cffi conflict

review refactor

review changes

Temp fixup

scheibelp commented on lib/spack/spack/bootstrap.py:
```diff
@@ -644,7 +644,7 @@ def _add_externals_if_missing():
         # clingo
         spack.repo.path.get('cmake'),
-        spack.repo.path.get('bison'),
+        # spack.repo.path.get('bison'),
```
This should be retained for non-Windows systems (and perhaps should search for `winbison` on Windows?)

scheibelp commented on lib/spack/spack/provider_index.py:
```diff
@@ -315,7 +315,6 @@ def mapiter(mappings):
             return six.iteritems(mappings)
         else:
             return iter(mappings)
-
```
This looks vestigial.

scheibelp commented on lib/spack/spack/spec.py:
```diff
@@ -1899,7 +1899,6 @@ def to_node_dict(self, hash=ht.dag_hash):
             hash (spack.hash_types.SpecHashDescriptor) type of hash to generate.
          """
         d = syaml.syaml_dict()
-
```
This looks vestigial.

scheibelp replied:
Ping (there are two files like this, including this file)

scheibelp commented on var/spack/repos/builtin/packages/re2c/package.py:
```diff
@@ -32,3 +32,10 @@ def configure_args(self):
         ]
+
+    def cmake(self, spec, prefix):
+        configure("--prefix="+ prefix, *self.configure_args())
```
If I understand correctly, there isn't actually a CMake build for `re2c` on Linux, but this achieves it because it replaces a call to `cmake` with a call to `configure`. I think while this representation is compact, it is potentially misleading. For now can we use the strategy of "demoting" to `Package`?

johnwparent replied:
Sounds good.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -55,11 +55,50 @@ def tar_support():
         any(path.endswith(t) for t in ALLOWED_ARCHIVE_TYPES)
 
 
+def lazy_caller(func_a, func_b, *callables):
```
I don't see this used anywhere.

johnwparent replied:
It's not. I wrote this before I realized that 7zip can atomically decompress/extract .txz files and implemented the rename strategy. If the renaming is acceptable, then I'll remove this.

scheibelp replied:
The renaming works, but if you prefer to use this and not rename, then that works too.

Is there documentation on any of the options/behavior of 7zip (i.e. did the documentation suggest changing the name to force extraction/unarchiving to occur together)?

johnwparent replied:
I'd prefer to just do the rename, so I'll remove this. I'm not sure if it's actually in the 7zip documentation, I just happened to noticed the `tar.xz` test was failing where the `.txz` test was not. 

Edit: After some searching I was unable to find a reference to this in the docs, so it seems like this may be undocumented behavior.

scheibelp replied:
As far as which approach is better, I should clarify: if there is documentation for one of the two methods (i.e. 7zip docs mention performing a separate step to untar) then I think that should be preferred. If both/neither are documented, then renaming seems good (although it would be worth adding a comment that this approach isn't documented).

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -235,7 +274,15 @@ def _7zip(archive_file):
     """
+    # 7zip will extract .txz atomically
+    # but not .tar.xz, rename tar.xz
+    # to .txz for atomic extraction
```
I understand what you mean here but I don't think "extract" is strictly defined to mean unarchiving the .tar, so can you clarify that in the comment?

johnwparent replied:
No problem.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -280,8 +327,19 @@ def decompressor_for(path, ext):
+            if txz_support():
+                return _untar
+            else:
+                # we only have garuntee of system support
```
Typo: guarantee

scheibelp commented on lib/spack/spack/build_systems/autotools.py:
```diff
@@ -104,7 +104,6 @@ def patch_config_files(self):
     depends_on('gnuconfig', type='build', when='target=ppc64le:')
     depends_on('gnuconfig', type='build', when='target=aarch64:')
     depends_on('gnuconfig', type='build', when='target=riscv64:')
-    conflicts('platform=windows')
```
(question) Can autotools packages now be installed on Windows?

scheibelp reviewed (changes requested):
I have some requests.

scheibelp commented on var/spack/repos/builtin/packages/re2c/package.py:
```diff
@@ -32,3 +45,21 @@ def configure_args(self):
+    @when('platform=windows')
+    def configure(self, spec, prefix):
+        with working_dir(self.stage.source_path, create=True):
+            args = ['-G', 'Ninja', '-DCMAKE_INSTALL_PREFIX=%s' % prefix]
```
This gives me the impression that this might depend on CMake on Windows; if so, there should be a `depends_on` declaration above.

scheibelp commented on var/spack/repos/builtin/packages/winbison/package.py:
```diff
@@ -0,0 +1,47 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
```
The header comments will need to be updated to 2022 instead of 2021.

scheibelp replied:
Ping on this

johnwparent replied:
Good catch.

scheibelp commented on var/spack/repos/builtin/packages/clingo/package.py:
```diff
@@ -45,14 +49,22 @@ class Clingo(CMakePackage):
+        depends_on('bison@2.5:', type="build", when='platform=cray')
+
+    with when('platform=windows'):
+        depends_on('re2c@0.13:', type="build")
```
(minor) It looks like you could merge this and the previous `when` block together (both add a dependency on `re2c` and the bison dependencies are already conditioned on version. My suggestion is wrong if in fact you need `winbison` to install point releases of Clingo on Windows (so I'm curious which version/s you have been installing so far).

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -235,7 +247,15 @@ def _7zip(archive_file):
+    # to .txz for atomic decompression
+    if 'tar.xz' in archive_file:
+        file_path = archive_file.strip('.tar.xz')
+        os.rename(archive_file, file_path + '.txz')
```
`archive_file` may refer to a reused file in the download cache, so I think we'd want to actually copy and rename this (otherwise we may rename the file inside of the cache).

scheibelp reviewed (changes requested):
Additional requests, all minor (this seems very close)

KiruyaMomochi commented:
I found that [the Clingo version used by spack](https://github.com/potassco/clingo/tree/2a025667090d71b2c9dce60fe924feb6bde8f667) failed to build with VS2022 and Python 3.10 due to lack of `ssize_t` and redefining `_DEBUG`. These problem can be resolved by 
- replace `ssize_t` with `Py_ssize_t`
- adding includes to `<yvals.h>` before undefining `_DEBUG` 

See https://github.com/KiruyaMomochi/clingo/pull/1 for the full diff.

johnwparent committed:
Proper library search for Python pkg

Fixup verion processing for Python

Remove print

Style updates

remove redundant spec context

Review

scheibelp commented on var/spack/repos/builtin/packages/clingo/package.py:
```diff
@@ -45,14 +49,22 @@ class Clingo(CMakePackage):
+
+    with when("platform=windows"):
+        depends_on("re2c@0.13:", type="build")
+        depends_on("winbison@2.4.12:", when="platform=windows")
```
Given the context manager, you can remove the `when` on this.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -234,7 +246,15 @@ def _7zip(archive_file):
+    # to .txz for atomic decompression
+    if "tar.xz" in archive_file:
+        file_path = archive_file.strip(".tar.xz")
+        os.rename(archive_file, file_path + ".txz")
```
A couple points:

* This has the same problem as https://github.com/spack/spack/pull/30690#discussion_r899382068. For now, the old file should be copied (or symlinked).
* `strip()` is a generally fragile when used for this: note that this looks like it would strip instances of the string ".tar.xz", but in fact ".tar.xz" is treated as a set of characters to strip - any order/combination of characters in that set will be removed

In all I recommend:

```
if ".tar.xz" in archive_file: #note the extra period at the beginning
  new_path = os.path.join(tmpdir, os.path.basename(archive_file.replace(".tar.gz", ".txz")))
  os.copy(archive_file, new_path)
  archive_file = new_path
```

johnwparent replied:
I think it makes more sense to fold this change into #31589 and rely on that PR prior to merging this one. The issue resolved by this (arguably hacky) workaround is the same as is being addressed in more recent changes in that PR.

scheibelp commented on lib/spack/spack/build_environment.py:
```diff
@@ -109,7 +109,8 @@
 
 # Platform-specific library suffix.
-dso_suffix = "dylib" if sys.platform == "darwin" else "so"
+dso_suffix = "dylib" if sys.platform == "darwin" else "dll" if sys.platform == "win32" else "so"
```
Could this be expanded into an if/else:

```
if sys.platform == "darwin":
  dso_suffix =
elif sys.platform == "win32":
  ...
else:
  dso_suffix = ".so"
```

johnwparent replied:
Amusingly enough, this change is actually from the RPath analog PR, not sure why it's flagged as a change here... Either way, will amend.

scheibelp commented on lib/spack/spack/util/compression.py:
```diff
@@ -285,8 +305,19 @@ def decompressor_for(path, ext):
+                # we only have guarantee of system support
+                return _7zip
+        if "Z" in ext:
+            return _7zip
```
If I understand, this isn't strictly needed: it looks like the change is to return `_untar` when possible in order to rely on native support when possible. Given that it falls back on `_7zip` - which is was always using before - I assume this particular change can be reverted. If that's true, I recommend rewording:

```
# Windows tar does not come with lzma or .Z support
# so we catch tarballs with those compressions and
# handle with 7zip if no native tar/lzma support
```

as

```
Use native decompression for .xz-compressed archives when possible.
```

johnwparent replied:
Like the other compression change, this is best handled in #31589 as that PR addresses this issue more directly.

scheibelp commented on var/spack/repos/builtin/packages/re2c/package.py:
```diff
@@ -22,6 +26,15 @@ class Re2c(AutotoolsPackage):
+    phases = ["configure", "build", "install"]
+
+    @property
+    def make(self):
```
Can this be called `make_tool` vs. `make`? I think ideally the name would suggest that it isn't always "make" that is being used.

johnwparent replied:
I'll be pulling this in favor of the re2c port PR, unless you think it'll be easier to merge if they are both part of the same PR.

scheibelp reviewed (changes requested):
I have some additional requests.

Also, now that the tests are running with Clingo, there appears to be an error effecting many of them where Spack cannot bootstrap Clingo:

```
>       raise ImportError(msg)
E       ImportError: cannot bootstrap the "clingo" Python module from spec "clingo-bootstrap@spack+python %msvc target=AMD64" due to the following failures:
E       github-actions raised ValueError: source is not trusted
E       github-actions-v0.2 raised ValueError: source is not trusted
E       github-actions-v0.1 raised ValueError: source is not trusted
E       spack-install raised AssertionError: RepoPath.get can only be called on concrete specs
```

scheibelp commented on lib/spack/spack/bootstrap.py:
```diff
@@ -655,6 +657,8 @@ def _add_externals_if_missing():
         spack.repo.path.get_pkg_class("gawk"),
     ]
+    if is_windows:
+        search_list.extend(spack.repo.path.get("winbison"))
```
To resolve the error mentioned in https://github.com/spack/spack/pull/30690#pullrequestreview-1108306432, replace `.get` with `.get_pkg_class`.

johnwparent committed:
Remove single line dso assignment

scheibelp commented on var/spack/repos/builtin/packages/clingo/package.py:
```diff
@@ -3,9 +3,13 @@
 from spack.compiler import UnsupportedCompilerFlag
 from spack.package import *
 
+is_windows = sys.platform == "win32"
```
This is no longer used in this package (I'm guessing it was added before spec-based constraints) and likewise for `import sys`.

johnwparent replied:
You are correct, and it has been removed.

scheibelp reviewed (changes requested):
Two minor issues:

* This PR added an import/sys.platform check that is no-longer needed (see comment)
* Address https://github.com/spack/spack/pull/30690/files#r971393931

Once those are dealt with this will be ready to go.

johnwparent committed:
Rename build tooling

scheibelp commented:
@spackbot run pipeline

johnwparent commented:
@spackbot run pipeline

KiruyaMomochi commented:
Bootstrap failed on my machine. Output of `spack -d spec hdf5`:

<details>

```
==> [2022-09-29-02:39:37.927942] Imported spec from built-in commands
==> [2022-09-29-02:39:37.943567] Imported spec from built-in commands
==> [2022-09-29-02:39:37.943567] Reading config from file C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\config.yaml
==> [2022-09-29-02:39:37.959194] Reading config from file C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\windows\config.yaml
==> [2022-09-29-02:39:37.959194] Skipping nonexistent config path \etc\spack\config.yaml
==> [2022-09-29-02:39:37.959194] Skipping nonexistent config path \etc\spack\windows\config.yaml
==> [2022-09-29-02:39:37.959194] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\config.yaml
==> [2022-09-29-02:39:37.959194] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\windows\config.yaml
==> [2022-09-29-02:39:37.959194] Reading config from file C:\Users\xtyzw\.spack\config.yaml
==> [2022-09-29-02:39:37.959194] Skipping nonexistent config path C:\Users\xtyzw\.spack\windows\config.yaml
==> [2022-09-29-02:39:37.959194] Reading config from file C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\bootstrap.yaml
==> [2022-09-29-02:39:37.959194] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\windows\bootstrap.yaml
==> [2022-09-29-02:39:37.959194] Skipping nonexistent config path \etc\spack\bootstrap.yaml
==> [2022-09-29-02:39:37.959194] Skipping nonexistent config path \etc\spack\windows\bootstrap.yaml
==> [2022-09-29-02:39:37.959194] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\bootstrap.yaml
==> [2022-09-29-02:39:37.959194] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\windows\bootstrap.yaml
==> [2022-09-29-02:39:37.959194] Skipping nonexistent config path C:\Users\xtyzw\.spack\bootstrap.yaml
==> [2022-09-29-02:39:37.959194] Skipping nonexistent config path C:\Users\xtyzw\.spack\windows\bootstrap.yaml
==> [2022-09-29-02:39:37.974823] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\upstreams.yaml
==> [2022-09-29-02:39:37.974823] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\windows\upstreams.yaml
==> [2022-09-29-02:39:37.974823] Skipping nonexistent config path \etc\spack\upstreams.yaml
==> [2022-09-29-02:39:37.978334] Skipping nonexistent config path \etc\spack\windows\upstreams.yaml
==> [2022-09-29-02:39:37.978334] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\upstreams.yaml
==> [2022-09-29-02:39:37.978334] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\windows\upstreams.yaml
==> [2022-09-29-02:39:37.978334] Skipping nonexistent config path C:\Users\xtyzw\.spack\upstreams.yaml
==> [2022-09-29-02:39:37.978334] Skipping nonexistent config path C:\Users\xtyzw\.spack\windows\upstreams.yaml
==> [2022-09-29-02:39:37.978334] Reading config from file C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\config.yaml
==> [2022-09-29-02:39:37.990849] Reading config from file C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\windows\config.yaml
==> [2022-09-29-02:39:37.990849] Skipping nonexistent config path \etc\spack\config.yaml
==> [2022-09-29-02:39:37.990849] Skipping nonexistent config path \etc\spack\windows\config.yaml
==> [2022-09-29-02:39:37.990849] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\config.yaml
==> [2022-09-29-02:39:37.990849] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\windows\config.yaml
==> [2022-09-29-02:39:37.990849] Reading config from file C:\Users\xtyzw\.spack\config.yaml
==> [2022-09-29-02:39:37.990849] Skipping nonexistent config path C:\Users\xtyzw\.spack\windows\config.yaml
==> [2022-09-29-02:39:37.990849] DATABASE LOCK TIMEOUT: 3s
==> [2022-09-29-02:39:37.990849] PACKAGE LOCK TIMEOUT: No timeout
==> [2022-09-29-02:39:37.990849] Reading config from file C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\bootstrap.yaml
==> [2022-09-29-02:39:38.006490] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\windows\bootstrap.yaml
==> [2022-09-29-02:39:38.006490] Skipping nonexistent config path \etc\spack\bootstrap.yaml
==> [2022-09-29-02:39:38.006490] Skipping nonexistent config path \etc\spack\windows\bootstrap.yaml
==> [2022-09-29-02:39:38.006490] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\bootstrap.yaml
==> [2022-09-29-02:39:38.006490] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\windows\bootstrap.yaml
==> [2022-09-29-02:39:38.006490] Skipping nonexistent config path C:\Users\xtyzw\.spack\bootstrap.yaml
==> [2022-09-29-02:39:38.006490] Skipping nonexistent config path C:\Users\xtyzw\.spack\windows\bootstrap.yaml
==> [2022-09-29-02:39:38.006490] DATABASE LOCK TIMEOUT: 3s
==> [2022-09-29-02:39:38.006490] PACKAGE LOCK TIMEOUT: No timeout
==> [2022-09-29-02:39:38.006490] [BOOTSTRAP CONFIG SCOPE] name=_builtin
==> [2022-09-29-02:39:38.006490] [BOOTSTRAP CONFIG SCOPE] name=defaults, path=C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults
==> [2022-09-29-02:39:38.006490] [BOOTSTRAP CONFIG SCOPE] name=defaults/windows, path=C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\windows
==> [2022-09-29-02:39:38.006490] [BOOTSTRAP CONFIG SCOPE] name=bootstrap, path=C:\Users\xtyzw\.spack\bootstrap\config
==> [2022-09-29-02:39:38.006490] [BOOTSTRAP CONFIG SCOPE] name=bootstrap/windows, path=C:\Users\xtyzw\.spack\bootstrap\config\windows
==> [2022-09-29-02:39:38.006490] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\compilers.yaml
==> [2022-09-29-02:39:38.006490] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\windows\compilers.yaml
==> [2022-09-29-02:39:38.006490] Skipping nonexistent config path C:\Users\xtyzw\.spack\bootstrap\config\compilers.yaml
==> [2022-09-29-02:39:38.006490] Reading config from file C:\Users\xtyzw\.spack\bootstrap\config\windows\compilers.yaml
==> [2022-09-29-02:39:38.697635] [BOOTSTRAP ROOT SPEC] clingo-bootstrap@spack+python %msvc target=AMD64
==> [2022-09-29-02:39:38.697635] [BOOTSTRAP MODULE clingo] Try importing from Python
==> [2022-09-29-02:39:38.697635] Reading config from file C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\bootstrap.yaml
==> [2022-09-29-02:39:38.697635] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\windows\bootstrap.yaml
==> [2022-09-29-02:39:38.697635] Skipping nonexistent config path C:\Users\xtyzw\.spack\bootstrap\config\bootstrap.yaml
==> [2022-09-29-02:39:38.697635] Bootstrapping clingo from pre-built binaries
==> [2022-09-29-02:39:38.713696] Reading config from file C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\config.yaml
==> [2022-09-29-02:39:38.714703] Reading config from file C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\windows\config.yaml
==> [2022-09-29-02:39:38.729213] Skipping nonexistent config path C:\Users\xtyzw\.spack\bootstrap\config\config.yaml
==> [2022-09-29-02:39:38.729213] DATABASE LOCK TIMEOUT: 3s
==> [2022-09-29-02:39:38.729213] PACKAGE LOCK TIMEOUT: No timeout
==> [2022-09-29-02:39:38.744859] Reading config from file C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\mirrors.yaml
==> [2022-09-29-02:39:38.744859] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\windows\mirrors.yaml
==> [2022-09-29-02:39:38.744859] Skipping nonexistent config path C:\Users\xtyzw\.spack\bootstrap\config\mirrors.yaml
==> [2022-09-29-02:39:38.744859] Skipping nonexistent config path C:\Users\xtyzw\.spack\bootstrap\config\windows\mirrors.yaml
==> [2022-09-29-02:39:38.744859] Checking existence of https://mirror.spack.io/bootstrap/github-actions/v0.3/build_cache/index.json
==> [2022-09-29-02:39:39.645459] Cached index for https://mirror.spack.io/bootstrap/github-actions/v0.3 already up to date
==> [2022-09-29-02:39:39.683164] Bootstrapping clingo from sources
==> [2022-09-29-02:39:39.803210] Reading config from file C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\config.yaml
==> [2022-09-29-02:39:39.818228] Reading config from file C:\Users\xtyzw\Projects\spack-2\etc\spack\defaults\windows\config.yaml
==> [2022-09-29-02:39:39.818228] Skipping nonexistent config path \etc\spack\config.yaml
==> [2022-09-29-02:39:39.818228] Skipping nonexistent config path \etc\spack\windows\config.yaml
==> [2022-09-29-02:39:39.818228] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\config.yaml
==> [2022-09-29-02:39:39.818228] Skipping nonexistent config path C:\Users\xtyzw\Projects\spack-2\etc\spack\windows\config.yaml
==> [2022-09-29-02:39:39.818228] Reading config from file C:\Users\xtyzw\.spack\config.yaml
==> [2022-09-29-02:39:39.818228] Skipping nonexistent config path C:\Users\xtyzw\.spack\windows\config.yaml
Traceback (most recent call last):
  File "C:\Users\xtyzw\Projects\spack-2\bin\spack", line 100, in <module>
    sys.exit(spack.main.main())
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\main.py", line 977, in main
    return _main(argv)
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\main.py", line 932, in _main
    return finish_parse_and_run(parser, cmd_name, env_format_error)
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\main.py", line 960, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\main.py", line 609, in _invoke_command
    return_val = command(parser, args)
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\cmd\spec.py", line 104, in spec
    concretized_specs = spack.cmd.parse_specs(args.specs, concretize=True)
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\cmd\__init__.py", line 227, in parse_specs
    spec.concretize(tests=tests)  # implies normalize
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\spec.py", line 2967, in concretize
    self._new_concretize(tests)
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\spec.py", line 2941, in _new_concretize
    solver = spack.solver.asp.Solver()
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\solver\asp.py", line 2319, in __init__
    self.driver = PyclingoDriver()
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\solver\asp.py", line 554, in __init__
    bootstrap_clingo()
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\solver\asp.py", line 534, in bootstrap_clingo
    spack.bootstrap.ensure_clingo_importable_or_raise()
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\bootstrap.py", line 778, in ensure_clingo_importable_or_raise
    ensure_module_importable_or_raise(module="clingo", abstract_spec=clingo_root_spec())
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\bootstrap.py", line 552, in ensure_module_importable_or_raise
    raise ImportError(msg)
ImportError: cannot bootstrap the "clingo" Python module from spec "clingo-bootstrap@spack+python %msvc target=AMD64" due to the following failures:
github-actions-v0.2 raised ValueError: source is not trusted
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\bootstrap.py", line 534, in ensure_module_importable_or_raise
    source_is_enabled_or_raise(current_config)
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\bootstrap.py", line 466, in source_is_enabled_or_raise
    raise ValueError("source is not trusted")

github-actions-v0.1 raised ValueError: source is not trusted
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\bootstrap.py", line 534, in ensure_module_importable_or_raise
    source_is_enabled_or_raise(current_config)
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\bootstrap.py", line 466, in source_is_enabled_or_raise
    raise ValueError("source is not trusted")

spack-install raised TypeError: 'PackageMeta' object is not iterable
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\bootstrap.py", line 537, in ensure_module_importable_or_raise
    if b.try_import(module, abstract_spec):
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\bootstrap.py", line 401, in try_import
    _add_externals_if_missing()
  File "C:\Users\xtyzw\Projects\spack-2\lib\spack\spack\bootstrap.py", line 661, in _add_externals_if_missing
    search_list.extend(spack.repo.path.get_pkg_class("winbison"))
```

</details>

scheibelp commented:
@KiruyaMomochi I spoke with @johnwparent about this earlier today and I think he is aware of how to fix both issues (he will be pushing a change soon)

alalazo commented:
I think this PR started making Windows unit tests fail. Apparently the CI job was still returning 0 so everything is :heavy_check_mark:

alalazo commented:
For reference, here's a link to the latest tests run on this PR:
- https://github.com/spack/spack/actions/runs/3136510560/jobs/5093623993

johnwparent commented:
Ok, thanks @alalazo if this is breaking things I have a good idea of what's going wrong. Fixing the "returns 0" issue should also be pretty straightforward.
